sys/net/gnrc/network_layer/ipv6/ext/frag/gnrc_ipv6_ext_frag.c

```diff
@@ -46,17 +60,256 @@ void gnrc_ipv6_ext_frag_init(void)
+    while (ptr) {
+        gnrc_pktsnip_t *tmp = gnrc_pktbuf_start_write(ptr);
+        if (tmp == NULL) {
+            DEBUG("ipv6_ext_frag: packet buffer full, canceling fragmentation\n");
+            if (ptr->users > 1) {
+                /* we are not the last fragment, so we need to also release
+                 * our hold on the snips we did not duplicate so far
+                 * and all also release all the snips we did duplicated so far
+                 */
+                if (to_send != NULL) {
+                    gnrc_pktbuf_release(to_send);
+                }
+                else {
+                    gnrc_pktbuf_release(ptr);
+                }
+            }
+            _snd_buf_free(snd_buf);
+            return;
+        }
+        ptr = tmp;
+        if (to_send == NULL) {
+            to_send = ptr;
+        }
+        switch (ptr->type) {
+            case GNRC_NETTYPE_IPV6: {
+                ipv6_hdr_t *hdr = ptr->data;
+                nh = &hdr->nh;
+                len = &hdr->len;
+                break;
+            }
+            case GNRC_NETTYPE_IPV6_EXT: {
+                ipv6_ext_t *hdr = ptr->data;
+                nh = &hdr->nh;
+                break;
+            }
+            default:
+                break;
+        }
+        if (ptr->type != GNRC_NETTYPE_NETIF) {
+            remaining -= ptr->size;
+        }
+        if (last) {
+            last->next = ptr;
+        }
+        last = ptr;
+        ptr = ptr->next;
+    }
+    assert(nh != NULL);
+    /* then the fragment header */
+    ptr = gnrc_ipv6_ext_build(last, last->next, *nh, sizeof(ipv6_ext_frag_t));
+    if (ptr == NULL) {
+        DEBUG("ipv6_ext_frag: unable to create fragmentation header\n");
+        gnrc_pktbuf_release(to_send);
+        _snd_buf_free(snd_buf);
+        return;
+    }
+    remaining -= sizeof(ipv6_ext_frag_t);
+    frag_hdr = ptr->data;
+    ipv6_ext_frag_set_offset(frag_hdr, snd_buf->offset);
+    if (!last_fragment) {
+        ipv6_ext_frag_set_more(frag_hdr);
+    }
+    frag_hdr->id = byteorder_htonl(snd_buf->id);
+    *nh = PROTNUM_IPV6_EXT_FRAG;
+    last = ptr;
+    /* then the rest */
+    while (remaining && snd_buf->pkt) {
+        if (last_fragment ||
+            (snd_buf->pkt->size <= remaining)) {
+            ptr = snd_buf->pkt;
+            snd_buf->pkt = ptr->next;
+            ptr->next = NULL;
+        }
+        else {
+            ptr = gnrc_pktbuf_mark(snd_buf->pkt, remaining,
+                                   GNRC_NETTYPE_UNDEF);
+            if (ptr == NULL) {
+                DEBUG("ipv6_ext_frag: packet buffer full, canceling fragmentation\n");
+                gnrc_pktbuf_release(to_send);
+                _snd_buf_free(snd_buf);
+                return;
+            }
+            assert(snd_buf->pkt->next == ptr);  /* we just created it with mark */
+            snd_buf->pkt->next = snd_buf->pkt->next->next;
+            ptr->next = NULL;
+        }
```

Comment:
You can move `ptr->next = NULL;` here, it's set in both cases.

Reply:
Done